libraries/libfc/include/fc/network/listener.hpp

```diff
@@ -58,30 +58,47 @@ struct listener_base<boost::asio::local::stream_protocol> {
 ///
+/// @tparam Protocol either \c boost::asio::ip::tcp or \c boost::asio::local::stream_protocol
+/// @tparam Executor The executor for the acceptor and acceptor timer
+/// @tparam SocketExecutorFn Function that returns executor to be used for the newly accepted socket:
+///                          ExecutorType(const endpoint_type&)
+/// @tparam CreateSession Callback Function to call after each accepted connection, provides newly created socket
+///                       and the Executor of the newly accepted socket: void(Socket&& socket, ExecutorType& ex)
+///                       Note that the ExecutorType provided by SocketExecutorFn can be retrieved from the socket
+///                       via socket.get_executor().
+///
 /////////////////////////////////////////////////////////////////////////////////////////////
-template <typename Protocol, typename Executor, typename CreateSession>
-struct listener : listener_base<Protocol>, std::enable_shared_from_this<listener<Protocol, Executor, CreateSession>> {
+template <typename Protocol, typename Executor, typename SocketExecutorFn, typename CreateSession>
+struct listener : listener_base<Protocol>, std::enable_shared_from_this<listener<Protocol, Executor, SocketExecutorFn, CreateSession>> {
  private:
    typename Protocol::acceptor      acceptor_;
    boost::asio::deadline_timer      accept_error_timer_;
    boost::posix_time::time_duration accept_timeout_;
    logger&                          logger_;
    std::string                      extra_listening_log_info_;
+   SocketExecutorFn                 socket_executor_fn_;
    CreateSession                    create_session_;
 
  public:
    using endpoint_type = typename Protocol::endpoint;
    listener(Executor& executor, logger& logger, boost::posix_time::time_duration accept_timeout,
             const std::string& local_address, const endpoint_type& endpoint,
-            const std::string& extra_listening_log_info, const CreateSession& create_session)
+            const std::string& extra_listening_log_info,
+            SocketExecutorFn socket_executor_fn, CreateSession create_session)
        : listener_base<Protocol>(local_address), acceptor_(executor, endpoint), accept_error_timer_(executor),
          accept_timeout_(accept_timeout), logger_(logger), extra_listening_log_info_(extra_listening_log_info),
-         create_session_(create_session) {}
+         socket_executor_fn_(std::move(socket_executor_fn)), create_session_(std::move(create_session)) {}
 
    const auto& acceptor() const { return acceptor_; }
 
    void do_accept() {
-      acceptor_.async_accept([self = this->shared_from_this()](boost::system::error_code ec, auto&& peer_socket) {
+      boost::system::error_code ec;
+      auto ep = acceptor_.local_endpoint(ec);
+      if (ec) {
+         fc_wlog(logger_, "Unable to retrieve local_endpoint of acceptor, error: ${e}", ("e", ec.message()));
+      }
+      acceptor_.async_accept(socket_executor_fn_(ep),
```

Comment:
What is the purpose for the changes here and for passing the local endpoint as a parameter to the `socket_executor_fn_`? (this local endpoint parameter is not used by any of ship/http/p2p)

Reply:
I thought it might be useful. If someone was wanting to choose which executor according to which endpoint. I'm happy to remove it since we don't have a current use-case.